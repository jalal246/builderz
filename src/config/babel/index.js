import babelPlugin from "./babelPlugin";
import { getPlugins, getPresets } from "./babelPresets";

export { babelPlugin, getPlugins, getPresets };
