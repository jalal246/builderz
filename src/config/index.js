module.exports = require("./rollup.config");
