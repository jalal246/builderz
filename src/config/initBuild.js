const packageSorter = require("package-sorter");

const {
  getPackagesPath,
  extractPackagesInfo,
  cleanBuildDir,
  msg
} = require("../utils");

/**
 * initBuild packages for production:
 *
 * 1- get path.
 * 2- validate and extract info.
 * 3- clean previous build.
 * 4- sort packages.
 *
 * @returns {Array} sorted packages
 */
function initBuild() {
  /**
   * get all packages.
   */
  const allPackages = getPackagesPath();

  /**
   * extract json form each package.
   */
  const packagesInfo = extractPackagesInfo({
    packages: allPackages,

    buildFileName: "dist",
    srcFileName: "src"
  });

  /**
   * Clean build if any.
   */
  cleanBuildDir({
    packages: allPackages,
    buildFilesName: ["dist"]
  });

  /**
   * Sort packages before bump to production.
   */
  const sorted = packageSorter(packagesInfo);

  msg("Done initiating build");

  return sorted;
}

module.exports = initBuild;
