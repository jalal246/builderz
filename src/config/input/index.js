module.exports = require("./getInput.js");
