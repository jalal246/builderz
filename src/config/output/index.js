module.exports = require("./getOutput.js");
