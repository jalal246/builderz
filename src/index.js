import { error } from "@mytools/print";
import start from "./rollup.config";

start().catch(err => {
  error(err);
});
