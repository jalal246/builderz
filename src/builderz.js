import { rollup } from "rollup";
import packageSorter from "package-sorter";
import { getJsonByName, getJsonByPath } from "get-info";

import { getInput, getOutput } from "./config/index";

import { isValidArr, bindFunc } from "./utils";

import { SORT_PACKAGES, PKG_PATHS, PKG_NAMES } from "./constants";

import StateHandler from "./store";

/**
 * Write bundle.
 *
 * @param {Object} inputOptions
 * @param {Object} outputOptions
 */
async function build(inputFunc, outputFunc, { isProd, buildFormat, order }) {
  try {
    const input = await inputFunc({
      isProd,
      buildFormat,
      idx: order,
    });

    const output = await outputFunc({
      isProd,
      buildFormat,
    });

    /**
     * create a bundle
     */
    const bundle = await rollup(input);

    /**
     * write the bundle to disk
     */
    await bundle.write(output);
  } catch (err) {
    console.error(err);
  }
}

async function bundlePkg(state, pkgInfo, json, index) {
  let pkgPath;

  state.setNewPkg();

  if (typeof json === "string") {
    pkgPath = json;
  } else {
    /**
     * When name isn't valid, get-info assign index-order instead of name.
     */
    ({ path: pkgPath } = pkgInfo[json.name || index]);

    state.assignJson(json);
  }

  state.checkOpts();

  await state.setPkgPath(pkgPath);
  const bundleOpt = state.unpackBundleOpts();

  state.extractName();
  state.extractEntries();
  state.extractAlias();

  const inputFunc = bindFunc(getInput, state);
  const outputFunc = bindFunc(getOutput, state);
  const buildFunc = bindFunc(build, inputFunc, outputFunc);

  await Promise.all(bundleOpt.map(buildFunc));
}

async function sortAndBump(pkgJson, bundlePkgFunc) {
  const { sorted, unSorted } = packageSorter(pkgJson);

  if (isValidArr(unSorted)) {
    console.error(`Unable to sort packages: ${unSorted}`);
  }

  await sorted.reduce(async (sortedPromise, json) => {
    await sortedPromise;

    await bundlePkgFunc(json);
  }, Promise.resolve());
}

async function builderz(opts) {
  const state = new StateHandler(opts);

  const {
    [PKG_NAMES]: pkgNames = [],
    [PKG_PATHS]: pkgPaths = [],
    [SORT_PACKAGES]: sortPackages = true,
  } = state.generalOpts;

  const { json: pkgJson, pkgInfo = {}, unfoundJson } = isValidArr(pkgNames)
    ? getJsonByName(...pkgNames)
    : getJsonByPath(...pkgPaths);

  const bundlePkgFunc = bindFunc(bundlePkg, state, pkgInfo);

  try {
    if (isValidArr(unfoundJson)) {
      await Promise.all(unfoundJson.map((path, i) => bundlePkgFunc(path, i)));
    } else if (isValidArr(pkgJson)) {
      const isSequence =
        pkgJson.length > 1 && !isValidArr(unfoundJson) && sortPackages;

      if (isSequence) {
        await sortAndBump(pkgJson, bundlePkgFunc);
      } else {
        await Promise.all(pkgJson.map(bundlePkgFunc));
      }
    }
  } catch (err) {
    console.error(err);
  }
}

export default builderz;
