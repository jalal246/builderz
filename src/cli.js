#!/usr/bin/env node

import error from "@mytools/print";
import builderz from "./builderz";
import resolveArgs from "./resolveArgs";

const globalArgs = resolveArgs();

function run() {
  try {
    builderz(globalArgs.opts(), { isInitOpts: false });
  } catch (err) {
    error(err);
  }
}

run();
