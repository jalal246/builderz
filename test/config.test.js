import { resolve } from "path";
import { initBuild } from "../src/config";

describe("config function", () => {
  it("initBuild default", () => {
    const { sorted, pkgInfo } = initBuild()();

    expect(sorted).toMatchSnapshot();
    expect(pkgInfo).toMatchSnapshot();
  });

  it("initBuild given path", () => {
    const packagesPath = resolve(__dirname, "packages-valid");

    const foloForms = resolve(packagesPath, "folo-forms");
    const foloLayout = resolve(packagesPath, "folo-layout");
    const foloUtils = resolve(packagesPath, "folo-utils");
    const foloWithcontext = resolve(packagesPath, "folo-withcontext");

    const { sorted, pkgInfo } = initBuild(
      "dist",
      foloForms,
      foloLayout,
      foloUtils,
      foloWithcontext
    )();

    expect(sorted).toMatchSnapshot();
    expect(pkgInfo).toMatchSnapshot();
  });
});
