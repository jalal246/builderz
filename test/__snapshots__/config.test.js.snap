// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config function initBuild default 1`] = `
Array [
  Object {
    "author": "Jalal Maskoun <jimmy002020@gmail.com>",
    "dependencies": Object {
      "@babel/core": "^7.8.3",
      "@babel/preset-env": "^7.8.3",
      "@mytools/print": "^0.1.3",
      "@rollup/plugin-auto-install": "^2.0.0",
      "@rollup/plugin-beep": "^0.1.2",
      "@rollup/plugin-commonjs": "^11.0.1",
      "@rollup/plugin-json": "^4.0.1",
      "@rollup/plugin-node-resolve": "^7.0.0",
      "camelize": "^1.0.0",
      "chalk": "^3.0.0",
      "commander": "^4.1.1",
      "del": "^5.1.0",
      "get-info": "^2.2.0",
      "glob": "^7.1.6",
      "ms": "^2.1.2",
      "package-sorter": "^1.1.1",
      "rimraf": "^3.0.0",
      "rollup": "^1.29.1",
      "rollup-plugin-analyzer": "^3.2.2",
      "rollup-plugin-babel": "^4.3.3",
      "rollup-plugin-terser": "^5.2.0",
    },
    "devDependencies": Object {
      "@babel/register": "^7.8.3",
      "babel-eslint": "^10.0.3",
      "chai": "^4.2.0",
      "eslint": "^6.8.0",
      "eslint-config-airbnb-base": "^14.0.0",
      "eslint-config-prettier": "^6.9.0",
      "eslint-plugin-import": "^2.20.0",
      "jest": "^25.1.0",
      "mocha": "^7.0.1",
    },
    "files": Array [
      "src",
      "dist",
      "README.md",
    ],
    "keywords": Array [
      "builderz",
      "build",
      "scripts",
      "monorepo",
      "rollup",
      "production",
      "utility",
      "tools",
      "package-sorter",
    ],
    "license": "LGPL-3.0",
    "main": "./dist/builderz.cjs",
    "name": "builderz",
    "peerDependencies": undefined,
    "publishConfig": Object {
      "access": "public",
      "registry": "https://registry.npmjs.org/",
    },
    "repository": "https://github.com/jalal246/builderz.git",
    "scripts": Object {
      "args": "node src -h",
      "build": "rollup src/rollup.config.js  --file dist/builderz.cjs.js --format=cjs",
      "prepublish": "yarn build",
      "test": "jest",
      "test2": "mocha -require @babel/register",
    },
    "sourcePath": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\src\\\\index.js",
    "version": "0.2.2",
  },
]
`;

exports[`config function initBuild default 2`] = `
Object {
  "builderz": Object {
    "dist": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\dist",
    "ext": "js",
  },
}
`;

exports[`config function initBuild given path 1`] = `
Array [
  Object {
    "author": "Jalal Maskoun <jimmy002020@gmail.com>",
    "bugs": Object {
      "url": "https://github.com/jalal246/folo/issues",
    },
    "dependencies": undefined,
    "description": "utils used in folo packages",
    "files": Array [
      "src",
      "dist",
      "README.md",
    ],
    "homepage": "https://github.com/jalal246/folo",
    "keywords": Array [
      "@folo",
      "@folo/layout",
      "@folo/values",
      "@folo/forms",
      "@folo/withcontext",
      "react",
      "react-component",
      "utility",
      "toolkit",
    ],
    "license": "MIT",
    "main": "dist/foloUtils.cjs.js",
    "module": "dist/foloUtils.esm.js",
    "name": "@folo/utils",
    "peerDependencies": undefined,
    "publishConfig": Object {
      "access": "public",
    },
    "repository": "https://github.com/jalal246/folo/tree/master/packages/folo-utils",
    "scripts": Object {},
    "sourcePath": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\test\\\\packages-valid\\\\folo-utils\\\\src\\\\index.js",
    "umd:main": "dist/foloUtils.min.umd.js",
    "version": "0.1.3",
  },
  Object {
    "author": "Jalal Maskoun <jimmy002020@gmail.com>",
    "bugs": Object {
      "url": "https://github.com/jalal246/folo/issues",
    },
    "dependencies": undefined,
    "description": "micro HOC compose component accepts custom context values as props",
    "files": Array [
      "src",
      "dist",
      "README.md",
    ],
    "homepage": "https://github.com/jalal246/folo",
    "keywords": Array [
      "@folo",
      "@folo/forms",
      "@folo/withcontext",
      "@folo/values",
      "@folo/layout",
      "@folo/utils",
      "react",
      "react-component",
      "context",
      "HOC",
      "utility",
      "higher-order",
      "toolkit",
    ],
    "license": "MIT",
    "main": "dist/foloWithcontext.cjs.js",
    "module": "dist/foloWithcontext.esm.js",
    "name": "@folo/withcontext",
    "peerDependencies": Object {
      "react": "^16.4.1",
    },
    "publishConfig": Object {
      "access": "public",
    },
    "repository": "https://github.com/jalal246/folo/tree/master/packages/folo-withcontext",
    "scripts": Object {},
    "sourcePath": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\test\\\\packages-valid\\\\folo-withcontext\\\\src\\\\index.js",
    "umd:main": "dist/foloWithcontext.min.umd.js",
    "version": "0.1.5",
  },
  Object {
    "author": "Jalal Maskoun <jimmy002020@gmail.com>",
    "bugs": Object {
      "url": "https://github.com/jalal246/folo/issues",
    },
    "dependencies": Object {
      "@folo/withcontext": "^0.1.5",
    },
    "description": "micro css grid react-components",
    "files": Array [
      "src",
      "dist",
      "README.md",
    ],
    "homepage": "https://github.com/jalal246/folo",
    "keywords": Array [
      "@folo",
      "@folo/forms",
      "@folo/withcontext",
      "@folo/values",
      "@folo/layout",
      "@folo/utils",
      "react",
      "react-component",
      "context",
      "css-grid",
      "css-layout",
      "form-builder-api",
      "dynamic-forms",
    ],
    "license": "MIT",
    "main": "dist/foloLayout.cjs.js",
    "module": "dist/foloLayout.esm.js",
    "name": "@folo/layout",
    "peerDependencies": Object {
      "prop-types": "^15.6.2",
      "react": "^16.8.0",
    },
    "publishConfig": Object {
      "access": "public",
    },
    "repository": "https://github.com/jalal246/folo/tree/master/packages/folo-layout",
    "scripts": Object {},
    "sourcePath": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\test\\\\packages-valid\\\\folo-layout\\\\src\\\\index.js",
    "umd:main": "dist/foloLayout.min.umd.js",
    "version": "0.1.4",
  },
  Object {
    "author": "Jalal Maskoun <jimmy002020@gmail.com>",
    "bugs": Object {
      "url": "https://github.com/jalal246/folo/issues",
    },
    "dependencies": Object {
      "@folo/layout": "^0.1.4",
      "@folo/values": "^0.1.4",
    },
    "description": "Full package combined of @folo/layout & @folo/values",
    "files": Array [
      "src",
      "dist",
      "README.md",
    ],
    "homepage": "https://github.com/jalal246/folo",
    "keywords": Array [
      "@folo",
      "@folo/forms",
      "@folo/withcontext",
      "@folo/values",
      "@folo/layout",
      "@folo/utils",
      "react",
      "react-component",
      "context",
      "css-grid",
      "css-layout",
      "form-builder-api",
      "dynamic-forms",
    ],
    "license": "MIT",
    "main": "dist/foloForms.cjs.js",
    "module": "dist/foloForms.esm.js",
    "name": "@folo/forms",
    "peerDependencies": Object {
      "prop-types": "^15.6.2",
      "react": "^16.8.0",
    },
    "publishConfig": Object {
      "access": "public",
    },
    "repository": "https://github.com/jalal246/folo/tree/master/packages/folo-forms",
    "scripts": Object {},
    "sourcePath": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\test\\\\packages-valid\\\\folo-forms\\\\src\\\\index.js",
    "umd:main": "dist/foloForms.min.umd.js",
    "version": "0.1.4",
  },
]
`;

exports[`config function initBuild given path 2`] = `
Object {
  "@folo/forms": Object {
    "dist": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\test\\\\packages-valid\\\\folo-forms\\\\dist",
    "ext": "js",
  },
  "@folo/layout": Object {
    "dist": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\test\\\\packages-valid\\\\folo-layout\\\\dist",
    "ext": "js",
  },
  "@folo/utils": Object {
    "dist": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\test\\\\packages-valid\\\\folo-utils\\\\dist",
    "ext": "js",
  },
  "@folo/withcontext": Object {
    "dist": "D:\\\\projects\\\\js\\\\published\\\\builderz\\\\test\\\\packages-valid\\\\folo-withcontext\\\\dist",
    "ext": "js",
  },
}
`;
